Daily long-run unit test failed in `TestParallelLockNewTask`

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
make ut-long
```

### 2. What did you expect to see? (Required)

Test passed.

### 3. What did you see instead (Required)

Test failed.

```
[2025/05/19 16:57:17.196 +08:00] [INFO] [task_manager_integration_test.go:116] ["lock new task with error"] [error="task is already scheduled"] [errorVerbose="task is already scheduled\ngithub.com/pingcap/tidb/pkg/ttl/ttlworker.init\n\t<autogenerated>:1\nruntime.doInit1\n\t/usr/local/go/src/runtime/proc.go:7350\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:7317\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:254\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1223\ngithub.com/pingcap/tidb/pkg/ttl/ttlworker.(*taskManager).lockScanTask.func1\n\t/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager.go:421\ngithub.com/pingcap/tidb/pkg/ttl/session.(*session).RunInTxn\n\t/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/session/session.go:169\ngithub.com/pingcap/tidb/pkg/ttl/ttlworker.(*taskManager).lockScanTask\n\t/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager.go:408\ngithub.com/pingcap/tidb/pkg/ttl/ttlworker.(*taskManager).LockScanTask\n\t/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager_test.go:47\ngithub.com/pingcap/tidb/pkg/ttl/ttlworker_test.TestParallelLockNewTask.func1\n\t/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager_integration_test.go:108\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_arm64.s:1223"]
panic: assert failed

goroutine 99378 [running]:
github.com/pingcap/tidb/pkg/util/intest.doPanic(...)
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/util/intest/assert_common.go:58
github.com/pingcap/tidb/pkg/util/intest.doAssert(...)
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/util/intest/assert_common.go:30
github.com/pingcap/tidb/pkg/util/intest.Assert(...)
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/util/intest/assert.go:25
github.com/pingcap/tidb/pkg/ttl/ttlworker.(*taskManager).lockScanTask.func1()
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager.go:450 +0xfe4
github.com/pingcap/tidb/pkg/ttl/session.(*session).RunInTxn(0xc0260fec60, {0x1096068d0, 0xc02633e6f0}, 0xc0248f83c0, 0x1)
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/session/session.go:169 +0x2f8
github.com/pingcap/tidb/pkg/ttl/ttlworker.(*taskManager).lockScanTask(0xc0248cec80, {0x109630f98, 0xc0260fec60}, 0xc02477b950, {0xc02633e5a0?, 0xc023ca09f0?, 0xc000258f50?})
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager.go:408 +0x31c
github.com/pingcap/tidb/pkg/ttl/ttlworker.(*taskManager).LockScanTask(...)
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager_test.go:47
github.com/pingcap/tidb/pkg/ttl/ttlworker_test.TestParallelLockNewTask.func1()
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager_integration_test.go:108 +0x300
created by github.com/pingcap/tidb/pkg/ttl/ttlworker_test.TestParallelLockNewTask in goroutine 87
	/Users/yangkeao/Projects/github.com/YangKeao/tidb/master/pkg/ttl/ttlworker/task_manager_integration_test.go:100 +0xa3c
^Cmake: *** [ut-long] Interrupt: 2
```

### 4. What is your TiDB version? (Required)

```
2dce92429d784d75c94146638c3ac1e437e859c8
```

Thanks for @purelind to report this issue 👍

--- a/pkg/session/syssession/session_test_util.go
+++ b/pkg/session/syssession/session_test_util.go
@@ -55,3 +55,9 @@ func (s *Session) IsInternalClosed() bool {
 func (p *AdvancedSessionPool) Size() int {
 	return len(p.pool)
 }
+
+// IsAvoidReuse returns whether the internal session avoids re-use
+// It is only used in testing
+func (s *Session) IsAvoidReuse() bool {
+	return s.internal.IsAvoidReuse()
+}

--- a/pkg/timer/tablestore/sql_test.go
+++ b/pkg/timer/tablestore/sql_test.go
@@ -709,7 +709,7 @@ func TestWithSession(t *testing.T) {
 	sctx.AssertExpectations(t)
 	mockCb1.AssertExpectations(t)
 
-	// rollback in restore failed, should close session
+	// rollback in restore failed, should avoid re-use session
 	mockSuccessInit()
 	mockCb1.On("cb1", pool.se).Return(nil).Once()
 	require.NoError(t, core.withSession(func(se *syssession.Session) error {
@@ -717,16 +717,15 @@ func TestWithSession(t *testing.T) {
 		sctx.On("ExecuteInternal", matchCtx, "ROLLBACK", []any(nil)).
 			Return(nil, errors.New("ROLLBACK error")).
 			Once()
-		sctx.On("Close").Once()
 		mockCb1.MethodCalled("cb1", se)
 		return nil
 	}))
-	require.True(t, pool.se.IsInternalClosed())
+	require.True(t, pool.se.IsAvoidReuse())
 	sctx.AssertExpectations(t)
 	mockCb1.AssertExpectations(t)
 	resetSe()
 
-	// set timezone in restore failed should close session
+	// set timezone in restore failed should avoid re-use
 	mockSuccessInit()
 	mockCb1.On("cb1", pool.se).Return(nil).Once()
 	require.NoError(t, core.withSession(func(se *syssession.Session) error {
@@ -737,11 +736,10 @@ func TestWithSession(t *testing.T) {
 		sctx.On("ExecuteInternal", matchCtx, "SET @@time_zone=%?", mock.Anything).
 			Return(nil, errors.New("SET tz error")).
 			Once()
-		sctx.On("Close").Once()
 		mockCb1.MethodCalled("cb1", se)
 		return nil
 	}))
-	require.True(t, pool.se.IsInternalClosed())
+	require.True(t, pool.se.IsAvoidReuse())
 	sctx.AssertExpectations(t)
 	mockCb1.AssertExpectations(t)
 	resetSe()

--- a/pkg/timer/tablestore/store.go
+++ b/pkg/timer/tablestore/store.go
@@ -375,13 +375,13 @@ func (s *tableTimerStoreCore) withSession(fn func(*syssession.Session) error) er
 				// Though `pool.WithSession` will discard a not committed transaction.
 				// We still rollback back here to make sure the assertion passes in `Pool.Put`.
 				terror.Log(err)
-				se.Close()
+				se.AvoidReuse()
 				return
 			}
 
 			if _, err = executeSQL(ctx, se, "SET @@time_zone=%?", originalTimeZone); err != nil {
 				terror.Log(err)
-				se.Close()
+				se.AvoidReuse()
 				return
 			}
 		}()

--- a/pkg/ttl/ttlworker/job_manager_integration_test.go
+++ b/pkg/ttl/ttlworker/job_manager_integration_test.go
@@ -58,6 +58,10 @@ import (
 )
 
 func sessionFactory(t *testing.T, from any) func() (session.Session, func()) {
+	return sessionFactoryWithTimeout(t, from, time.Minute)
+}
+
+func sessionFactoryWithTimeout(t *testing.T, from any, timeout time.Duration) func() (session.Session, func()) {
 	var pool syssession.Pool
 	switch p := from.(type) {
 	case *domain.Domain:
@@ -80,7 +84,7 @@ func sessionFactory(t *testing.T, from any) func() (session.Session, func()) {
 
 				select {
 				case <-ch:
-				case <-time.After(time.Minute):
+				case <-time.After(timeout):
 					require.FailNow(t, "timeout")
 				}
 				return nil
@@ -183,7 +187,11 @@ func TestParallelLockNewJob(t *testing.T) {
 	waitAndStopTTLManager(t, dom)
 	tk := testkit.NewTestKit(t, store)
 
-	sessionFactory := sessionFactory(t, dom)
+	sessionTimeout := time.Minute
+	if testflag.Long() {
+		sessionTimeout = 10 * time.Minute
+	}
+	sessionFactory := sessionFactoryWithTimeout(t, dom, sessionTimeout)
 
 	testTable := &cache.PhysicalTable{ID: 2, TableInfo: &model.TableInfo{ID: 1, TTLInfo: &model.TTLInfo{IntervalExprStr: "1", IntervalTimeUnit: int(ast.TimeUnitDay), JobInterval: "1h"}}}
 	// simply lock a new job
@@ -1830,7 +1838,6 @@ func overwriteJobInterval(t *testing.T) func() {
 }
 
 func TestJobManagerWithFault(t *testing.T) {
-	// TODO: add a flag `-long` to enable this test
 	skip.NotUnderLong(t)
 
 	defer boostJobScheduleForTest(t)()

--- a/pkg/ttl/ttlworker/task_manager_integration_test.go
+++ b/pkg/ttl/ttlworker/task_manager_integration_test.go
@@ -50,7 +50,11 @@ func TestParallelLockNewTask(t *testing.T) {
 	testTable, err := tk.Session().GetDomainInfoSchema().(infoschema.InfoSchema).TableByName(context.Background(), ast.NewCIStr("test"), ast.NewCIStr("t"))
 	require.NoError(t, err)
 
-	sessionFactory := sessionFactory(t, dom)
+	sessionTimeout := time.Minute
+	if testflag.Long() {
+		sessionTimeout = 10 * time.Minute
+	}
+	sessionFactory := sessionFactoryWithTimeout(t, dom, sessionTimeout)
 	se, closeSe := sessionFactory()
 	defer closeSe()
 

